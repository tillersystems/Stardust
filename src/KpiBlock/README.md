# Kpi Block

### Usage

```jsx
import { KpiBlock } from '@tillersystems/stardust';
```

A Kpi Block provides useful data pieces of information that should help users identity
the most important things about their data.

<!-- STORY -->

### Properties

| propName    | Required |     propType      | defaultValue |                                   Description                                   |
| ----------- | :------: | :---------------: | :----------: | :-----------------------------------------------------------------------------: |
| `className` |    -     |     `string`      |     `''`     |                        Class needed by styled component                         |
| `title`     |    \*    |     `string`      |              |                         Title displayed below the value                         |
| `value`     |    \*    | `string` / `node` |              |                   Value displayed at the center of the block                    |
| `variation` |    -     | `number` / `bool` |              | Variation, positive or negative, displayed at the top right corner of the block |

### Example

```jsx
import { KpiBlock } from '@tillersystems/stardust';

render() {
  return <KpiBlock title="commandes" value="657" variation={5} />
}
```
