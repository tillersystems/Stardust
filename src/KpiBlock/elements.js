import styled, { css } from 'styled-components';

export const Value = styled.div`
  font-size: ${({ theme: { fonts } }) => fonts.size.big};
  font-weight: ${({ theme: { fonts } }) => fonts.weight.thick};

  color: ${({ theme: { palette } }) => palette.darkBlue};
`;

export const Title = styled.div`
  font-size: ${({ theme: { fonts } }) => fonts.size.medium};

  color: ${({ theme: { palette } }) => palette.mediumGrey};
`;

export const Variation = styled.div`
  position: absolute;
  top: 1rem;
  right: 1rem;

  display: flex;
  align-items: center;

  font-size: ${({ theme: { fonts } }) => fonts.size.default};

  /* Negative */
  ${({ negative }) =>
    negative &&
    css`
      color: ${({ theme: { palette } }) => palette.failure.default};
    `};

  /* Positive */
  ${({ positive }) =>
    positive &&
    css`
      color: ${({ theme: { palette } }) => palette.success.default};
    `};
`;
