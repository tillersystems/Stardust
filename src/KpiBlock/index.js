import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Icon, Theme } from '..';
import { Title, Value, Variation } from './elements';

/**
 * Kpi Block
 *
 * This component is in charge of displaying
 * a Kpi Block
 *
 * @param {string} className // Class needed by styled component.
 * @param {string} title // KpiBlock title.
 * @param {string} value // KpiBlock value.
 * @param {number} variation // KpiBlock variation (negative or positive).
 *
 * @return {jsx}
 */

const KpiBlock = ({ className, title, value, variation }) => (
  <div className={className}>
    {variation !== false && variation < 0 && (
      <Variation negative>
        {variation.toFixed(2)} % <Icon name="caret-down" color={Theme.palette.failure.default} />
      </Variation>
    )}
    {variation !== false && variation >= 0 && (
      <Variation positive>
        +{variation.toFixed(2)} % <Icon name="caret-up" color={Theme.palette.success.default} />
      </Variation>
    )}
    {typeof value === 'string' && <Value>{value}</Value>}
    {React.isValidElement(value) && value}
    <Title>{title}</Title>
  </div>
);

/**
 * PropTypes Validation
 */
const { bool, number, node, oneOfType, string } = PropTypes;
KpiBlock.propTypes = {
  className: string,
  title: string.isRequired,
  value: oneOfType([string, node]).isRequired,
  variation: oneOfType([bool, number]),
};

/**
 * Default props
 */
KpiBlock.defaultProps = {
  className: '',
  variation: false,
};

export default styled(KpiBlock)`
  align-items: center;
  background: ${({ theme: { palette } }) => palette.white};
  border: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
  border-radius: ${({ theme: { dimensions } }) => dimensions.radius};
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.04);
  box-sizing: content-box;
  display: flex;
  flex-direction: column;
  height: 6.6rem;
  padding: 3rem 1rem;
  position: relative;
`;
