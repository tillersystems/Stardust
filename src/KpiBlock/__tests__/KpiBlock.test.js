import React from 'react';
import 'jest-styled-components';

import KpiBlock from '..';

describe('<KpiBlock />', () => {
  it('should render withouth a problem', () => {
    const render = mountWithTheme(<KpiBlock title="title" value="value" />);

    expect(render).toMatchSnapshot();
  });

  it('should render with a positive variation value', () => {
    const render = mountWithTheme(<KpiBlock title="title" value="value" variation={100} />);

    expect(render).toMatchSnapshot();
  });

  it('should render with a negative variation value', () => {
    const render = mountWithTheme(<KpiBlock title="title" value="value" variation={-100} />);

    expect(render).toMatchSnapshot();
  });

  it('should render with a custom a value', () => {
    const render = mountWithTheme(
      <KpiBlock title="title" value="<div>67</div>" variation={-100} />,
    );

    expect(render).toMatchSnapshot();
  });
});
