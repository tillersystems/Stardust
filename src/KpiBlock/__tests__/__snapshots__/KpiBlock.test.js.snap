// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KpiBlock /> should render with a negative variation value 1`] = `
.c2 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c3 {
  color: hsl(213,17%,20%);
  font-size: 3rem;
  font-weight: 600;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.2rem;
  position: absolute;
  right: 1rem;
  top: 1rem;
  color: hsl(6,79%,65%);
}

.c4 {
  color: hsl(207,13%,45%);
  font-size: 1.4rem;
  margin-top: 0.5rem;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: hsl(0,100%,100%);
  border: 1px solid hsl(207,22%,90%);
  border-radius: 0.4rem;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.04);
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 6.6rem;
  padding: 3rem 1rem;
  position: relative;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    -100.00
     % 
    <div
      class="c2"
      data-testid="iconContainer"
      height="20"
      width="20"
    >
      <svg
        data-testid="iconSvg"
        height="20"
        style="display: block;"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m 384.662 160 h -257.324 c -17.818 0 -26.741 21.543 -14.142 34.142 l 128.662 128.662 c 7.81 7.81 20.474 7.81 28.284 0 l 128.662 -128.662 c 12.6 -12.599 3.676 -34.142 -14.142 -34.142 z"
          data-testid="iconSvgPath"
          fill="hsl(6,79%,65%)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="c3"
  >
    value
  </div>
  <div
    class="c4"
  >
    title
  </div>
</div>
`;

exports[`<KpiBlock /> should render with a positive variation value 1`] = `
.c2 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c3 {
  color: hsl(213,17%,20%);
  font-size: 3rem;
  font-weight: 600;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.2rem;
  position: absolute;
  right: 1rem;
  top: 1rem;
  color: hsl(89,53%,52%);
}

.c4 {
  color: hsl(207,13%,45%);
  font-size: 1.4rem;
  margin-top: 0.5rem;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: hsl(0,100%,100%);
  border: 1px solid hsl(207,22%,90%);
  border-radius: 0.4rem;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.04);
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 6.6rem;
  padding: 3rem 1rem;
  position: relative;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    +
    100.00
     % 
    <div
      class="c2"
      data-testid="iconContainer"
      height="20"
      width="20"
    >
      <svg
        data-testid="iconSvg"
        height="20"
        style="display: block;"
        viewBox="0 0 512 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m 384.662 352 h -257.324 c -17.818 0 -26.741 -21.543 -14.142 -34.142 l 128.662 -128.662 c 7.81 -7.81 20.474 -7.81 28.284 0 l 128.662 128.662 c 12.6 12.599 3.676 34.142 -14.142 34.142 z"
          data-testid="iconSvgPath"
          fill="hsl(89,53%,52%)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
  <div
    class="c3"
  >
    value
  </div>
  <div
    class="c4"
  >
    title
  </div>
</div>
`;

exports[`<KpiBlock /> should render without a problem 1`] = `
.c1 {
  color: hsl(213,17%,20%);
  font-size: 3rem;
  font-weight: 600;
}

.c2 {
  color: hsl(207,13%,45%);
  font-size: 1.4rem;
  margin-top: 0.5rem;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: hsl(0,100%,100%);
  border: 1px solid hsl(207,22%,90%);
  border-radius: 0.4rem;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.04);
  box-sizing: content-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 6.6rem;
  padding: 3rem 1rem;
  position: relative;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    value
  </div>
  <div
    class="c2"
  >
    title
  </div>
</div>
`;
